### PR TITLE
IDGEN-33: Change db setup to use liquibase

### PR DESCRIPTION
**Pull request by segunolalive**
## JIRA TICKET NAME
[IDGEN-33: Change db setup to use liquibase](https://issues.openmrs.org/browse/IDGEN-33)

### SUMMARY

Take the existing sqldiff scripts and migrate them to liquibase. Make sure the liquibase scripts work on postgres.

Upgrade OpenMRS minimum version dependency to 1.6.6 for liquibase support.

This PR migrates sqldiffs to liquibase version 2.0

### CODE REVIEW

**Comment by dkayiwa** [comment]
@segunolalive you should create a liquibase.xml file in this folder: https://github.com/openmrs/openmrs-module-idgen/tree/master/api/src/main/resources

That is where all the changesets should be. And it is the only change you need to make for this ticket. This is exactly why i encourage you to commit early and often (daily) such that we guide you before you go too far in the wrong direction.

**Comment by dkayiwa** [comment]
Did you also take a look at this? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by segunolalive** [comment]
Okay. Thanks

**Comment by dkayiwa** [comment]
@segunolalive i would like to start by asking that did you actually read this? https://wiki.openmrs.org/display/docs/Pull+Request+Tips

**Comment by segunolalive** [comment]
Yes, I did.

Did I do something wrong?

**Comment by dkayiwa** [comment]
Take a second look and tell me if you did not skip anything.

**Comment by segunolalive** [comment]
I skipped the first instruction. I never assigned the ticket to myself.

**Comment by dkayiwa** [comment]
Can you please read them again one by one without skipping anything?

**Comment by segunolalive** [comment]
Done that.

**Comment by dkayiwa** [comment]
@segunolalive and you did not skip anything?

**Comment by segunolalive** [comment]
I don't think I have

**Comment by dkayiwa** [comment]
@segunolalive do you have some time now and setup a hangout call? Would like you to take me through each of those steps and how you have done it.

**Comment by segunolalive** [comment]
Hangout link:
https://hangouts.google.com/hangouts/_/calendar/c2VndW5vbGFsaXZlQGdtYWlsLmNvbQ.5a1i9prdgt7a0tckd845dgq3ob?authuser=1 <https://hangouts.google.com/hangouts/_/calendar/c2VndW5vbGFsaXZlQGdtYWlsLmNvbQ.5a1i9prdgt7a0tckd845dgq3ob?authuser=1>

**Comment by dkayiwa** [comment]
Did you see the comment that i just put on the JIRA ticket?

**Comment by segunolalive** [comment]
Seen. Thank you. Would look into it

**Comment by Charpell** [comment]
Looks good to me

File: api/src/main/resources/liquibase.xml [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+<databaseChangeLog logicalFilePath="liquibase-update-to-latest.xml" xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd
+    http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd">
+
+
+	<changeSet id="1.0" author="Partners In Health">
+		<preConditions onError="MARK_RAN" onFail="MARK_RAN">
+			<not><tableExists tableName="idgen_identifier_source" /></not>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Each table creation should have its own changeset.

Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: okay

File: api/src/main/resources/liquibase.xml [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+		xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/2.0
+         http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd"> -->
+
+<databaseChangeLog logicalFilePath="liquibase-update-to-latest.xml" xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+	xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd
+    http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd">
+
+
+	<changeSet id="1.0" author="Partners In Health">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Can you also make the changeset ids to either be uuids or of a pattern like "idgen_identifier_source_table_create"?

Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: Should this pattern apply to other changesets as well, such as `idgen_auto_generation_option_add_foreign-key_constraint`?

Comment (dkayiwa) [reply api/src/main/resources/liquibase.xml]: The constraints should be contained in their table creation changesets

File: api/src/main/resources/liquibase.xml [hunk 3]
```diff
@@ -0,0 +1,329 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- <databaseChangeLog
+		xmlns="http://www.liquibase.org/xml/ns/dbchangelog/2.0"
+		xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+		xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/2.0
+         http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-2.0.xsd"> -->
+
+<databaseChangeLog logicalFilePath="liquibase-update-to-latest.xml" xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: We do not need logicalFilePath for modules.

Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: okay. I'll remove that

Comment (dkayiwa) [reply api/src/main/resources/liquibase.xml]: Have you also seen my comment on JIRA?

Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: I have. Just responded

File: api/src/main/resources/liquibase.xml [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+		<preConditions onFail="MARK_RAN">
+			<tableExists tableName="idgen_seq_id_gen" />
+			<not><columnExists tableName="idgen_seq_id_gen" columnName="max_length"/></not>
+		</preConditions>
+		<addColumn tableName="idgen_seq_id_gen">
+			<column name="max_length" type="int(11)">
+				<constraints nullable="true"/>
+			</column>
+		</addColumn>
+		<sql>UPDATE `idgen_seq_id_gen` SET `max_length` = `min_length`</sql>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: The above sql update statement should be in its own changeset.

File: api/src/main/resources/liquibase.xml [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+			<column name="uuid" type="char(38)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="name" type="varchar(255)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="description" type="varchar(1000)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="identifier_type" type="int(11)" defaultValue="0">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Let us drop all default values of 0 for integer fields

File: api/src/main/resources/liquibase.xml [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+				<constraints nullable="false" unique="true" foreignKeyName="identifier_type for idgen_identifier_source" references="patient_identifier_type(patient_identifier_type_id)"/>
+			</column>
+			<column name="creator" type="int(11)" defaultValue="0">
+				<constraints nullable="false" unique="true" foreignKeyName="creator for idgen_identifier_source" references="users(user_id)"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Are you sure that the creator should be unique? Meaning that one cannot create more than one row in this table?

File: api/src/main/resources/liquibase.xml [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+			<column name="id" type="int(11)" autoIncrement="true">
+				<constraints primaryKey="true" nullable="false" unique="true"/>
+			</column>
+			<column name="uuid" type="char(38)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="name" type="varchar(255)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="description" type="varchar(1000)">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Are you sure that the description cannot be null?

File: api/src/main/resources/liquibase.xml [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+
+	<changeSet id="43889130-55c2-49e9-b72b-a9eae9a31da7" author="Partners In Health">
+		<preConditions onFail="MARK_RAN">
+			<not><tableExists tableName="idgen_identifier_source"/></not>
+		</preConditions>
+		<createTable tableName="idgen_identifier_source">
+			<column name="id" type="int(11)" autoIncrement="true">
+				<constraints primaryKey="true" nullable="false" unique="true"/>
+			</column>
+			<column name="uuid" type="char(38)">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: The uuid should be unique

File: api/src/main/resources/liquibase.xml [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+			<column name="creator" type="int(11)" defaultValue="0">
+				<constraints nullable="false" unique="true" foreignKeyName="creator for idgen_identifier_source" references="users(user_id)"/>
+			</column>
+			<column name="date_created" type="datetime">
+				<constraints nullable="false"/>
+			</column>
+			<column name="changed_by" type="int(11)" defaultValue="NULL">
+				<constraints nullable="true" unique="true" foreignKeyName="changed_by for idgen_identifier_source" references="users(user_id)"/>
+			</column>
+			<column name="date_changed" type="datetime" defaultValue="NULL">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Let us just drop the default attribute for this.

File: api/src/main/resources/liquibase.xml [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+				<constraints nullable="true"/>
+			</column>
+			<column name="retired" type="tinyint(1)" defaultValue="0">
+				<constraints nullable="false"/>
+			</column>
+			<column name="retired_by" type="int(11)" defaultValue="NULL">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Let us just drop the default attribute for this.

Comment (dkayiwa) [reply api/src/main/resources/liquibase.xml]: For the same reasons as creator, this should not be unique.

File: api/src/main/resources/liquibase.xml [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+				<constraints nullable="true" unique="true" foreignKeyName="retired_by for idgen_identifier_source" references="users(user_id)"/>
+			</column>
+			<column name="date_retired" type="datetime" defaultValue="NULL">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Let us just drop the default attribute for this.

File: api/src/main/resources/liquibase.xml [hunk 12]
```diff
@@ -0,0 +1,320 @@
...
+				<constraints nullable="true"/>
+			</column>
+			<column name="retire_reason" type="varchar(255)" defaultValue="NULL">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Let us just drop the default attribute for this.

File: api/src/main/resources/liquibase.xml [hunk 13]
```diff
@@ -0,0 +1,320 @@
...
+			<column name="identifier_type" type="int(11)" defaultValue="0">
+				<constraints nullable="false" unique="true" foreignKeyName="identifier_type for idgen_identifier_source" references="patient_identifier_type(patient_identifier_type_id)"/>
+			</column>
+			<column name="creator" type="int(11)" defaultValue="0">
+				<constraints nullable="false" unique="true" foreignKeyName="creator for idgen_identifier_source" references="users(user_id)"/>
+			</column>
+			<column name="date_created" type="datetime">
+				<constraints nullable="false"/>
+			</column>
+			<column name="changed_by" type="int(11)" defaultValue="NULL">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: This should not be unique

File: api/src/main/resources/liquibase.xml (hunk 6 above)
Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: Remove the unique constraint would mean, the table cannot be indexed by the creator column as specified in the sqldiff

File: api/src/main/resources/liquibase.xml (hunk 7 above)
Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: okay. I'll change that

File: api/src/main/resources/liquibase.xml (hunk 8 above)
Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: Alright. Thanks

File: api/src/main/resources/liquibase.xml (hunk 6 above)
Comment (dkayiwa) [reply api/src/main/resources/liquibase.xml]: Could you point me to where it is indexed by the creator column? As in which line number.

File: api/src/main/resources/liquibase.xml [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+				<constraints nullable="false"/>
+			</column>
+			<column name="base_character_set" type="varchar(255)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="first_identifier_base" type="varchar(50)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="prefix" type="varchar(20)">
+				<constraints nullable="false"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Are you sure this is not nullable?

File: api/src/main/resources/liquibase.xml [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+			</column>
+			<column name="suffix" type="varchar(20)">
+				<constraints nullable="false"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Are you sure this is not nullable?

File: api/src/main/resources/liquibase.xml [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+			</column>
+			<column name="length" type="int(11)">
+				<constraints nullable="false"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Are you sure this is not nullable?

File: api/src/main/resources/liquibase.xml (hunk 9 above)
Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: oaky.

File: api/src/main/resources/liquibase.xml [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+
+	<changeSet id="b50c8351-3c6c-40eb-be40-bbdaed78c5a0" author="Partners In Health">
+		<preConditions onFail="MARK_RAN">
+			<not><tableExists tableName="idgen_id_pool" /></not>
+		</preConditions>
+		<createTable tableName="idgen_id_pool">
+			<column name="id" type="int(11)">
+				<constraints nullable="false" primaryKey="true" foreignKeyName="id for idgen_id_pool" references="idgen_identifier_source(id)"/>
+			</column>
+			<column name="source" type="int(11)">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Is this not nullable and unique?

Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Is this unique?

File: api/src/main/resources/liquibase.xml [hunk 18]
```diff
@@ -0,0 +1,320 @@
...
+
+	<changeSet id="9b1eb03d-bd27-428b-8359-5877797b0fb8" author="Partners In Health">
+		<preConditions onFail="MARK_RAN">
+			<not><tableExists tableName="idgen_pooled_identifier" /></not>
+		</preConditions>
+		<createTable tableName="idgen_pooled_identifier">
+			<column name="id" type="int(11)" autoIncrement="true">
+				<constraints nullable="false" primaryKey="true"/>
+			</column>
+			<column name="uuid" type="char(38)">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: This should also be unique

File: api/src/main/resources/liquibase.xml (hunk 6 above)
Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: https://github.com/openmrs/openmrs-module-idgen/blob/448573a0d360a6929893d89c062f9968e5fbcf0e/omod/src/main/resources/sqldiff.xml#L32

File: api/src/main/resources/liquibase.xml [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+	<changeSet id="4b4ee08c-262f-48f0-a37e-2372614e1a58" author="Partners In Health">
+		<preConditions onFail="MARK_RAN">
+			<not><tableExists tableName="idgen_log_entry" /></not>
+		</preConditions>
+		<createTable tableName="idgen_log_entry">
+			<column name="id" type="int(11)" autoIncrement="true">
+				<constraints nullable="false" primaryKey="true" unique="true"/>
+			</column>
+			<column name="source" type="int(11)">
+				<constraints nullable="false" unique="true" foreignKeyName="source for idgen_log" references="idgen_identifier_source(id)"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Is this specified as unique in the sqldiff?

File: api/src/main/resources/liquibase.xml (hunk 6 above)
Comment (dkayiwa) [reply api/src/main/resources/liquibase.xml]: Does indexing a table field require its values to be unique?

File: api/src/main/resources/liquibase.xml [hunk 20]
```diff
@@ -0,0 +1,320 @@
...
+			</column>
+			<column name="identifier" type="varchar(50)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="date_generated" type="datetime">
+				<constraints nullable="false"/>
+			</column>
+			<column name="generated_by" type="int(11)">
+				<constraints nullable="false" unique="true" foreignKeyName="generated_by for idgen_log" references="users(user_id)"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Is this specified as unique in the sqldiff?

File: api/src/main/resources/liquibase.xml [hunk 21]
```diff
@@ -0,0 +1,320 @@
...
+			</column>
+			<column name="comment" type="varchar(255)" defaultValue="NULL">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Let us drop the default value attribute

File: api/src/main/resources/liquibase.xml [hunk 22]
```diff
@@ -0,0 +1,320 @@
...
+	<changeSet id="17a1d838-bf40-45d0-9e2b-6d177ff7ad62" author="Partners In Health">
+		<preConditions onFail="MARK_RAN">
+			<not><tableExists tableName="idgen_reserved_identifier" /></not>
+		</preConditions>
+		<createTable tableName="idgen_reserved_identifier">
+			<column name="id" type="int(11)" autoIncrement="true">
+				<constraints nullable="false" primaryKey="true" unique="true"/>
+			</column>
+			<column name="source" type="int(11)">
+				<constraints nullable="false" unique="true" foreignKeyName="source for idgen_reserved_identifier" references="idgen_identifier_source(id)"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Is this specified as unique in the sqldiff?

File: api/src/main/resources/liquibase.xml (hunk 17 above)
Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: It thought it should be because the column has a foreignKey constraint

File: api/src/main/resources/liquibase.xml [hunk 23]
```diff
@@ -0,0 +1,320 @@
...
+		</preConditions>
+		<addColumn tableName="idgen_remote_source">
+			<column name="password" type="varchar(20)">
+				<constraints nullable="true"/>
+			</column>
+		</addColumn>
+	</changeSet>
+
+	<changeSet id="aea80abc-8e1b-44f2-9574-c64741750cec" author="Jeremy Keiper">
+		<preConditions onFail="MARK_RAN">
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: I think in this case we do not need the first precondition because it is implied in the second precondition. Not so?

File: api/src/main/resources/liquibase.xml (hunk 17 above)
Comment (dkayiwa) [reply api/src/main/resources/liquibase.xml]: Are you sure foreign key constraint fields should be unique in the referencing table?

File: api/src/main/resources/liquibase.xml [hunk 24]
```diff
@@ -0,0 +1,320 @@
...
+		</preConditions>
+		<dropIndex tableName="idgen_auto_generation_option" indexName="identifier_type"/>
+ 	</changeSet>
+
+	<changeSet id="0f79a646-4f62-4fef-b035-421ad8f73ba2" author="Partners In Health">
+		<preConditions onFail="MARK_RAN">
+			<not><foreignKeyConstraintExists foreignKeyName="identifier_type" /></not>
+		</preConditions>
+		<addForeignKeyConstraint
+				constraintName="identifier_type for idgen_auto_generation_option"
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: The constraint name does not match the one in the precondition.

File: api/src/main/resources/liquibase.xml (hunk 5 above)
Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: Yes it does

File: api/src/main/resources/liquibase.xml [hunk 25]
```diff
@@ -0,0 +1,317 @@
...
+			</column>
+			<column name="uuid" type="char(38)">
+				<constraints nullable="false" unique="true"/>
+			</column>
+			<column name="name" type="varchar(255)">
+				<constraints nullable="false"/>
+			</column>
+			<column name="description" type="varchar(1000)" />
+			<column name="identifier_type" type="int(11)">
+				<constraints nullable="false" unique="true" foreignKeyName="identifier_type for idgen_identifier_source" references="patient_identifier_type(patient_identifier_type_id)"/>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: Is this unique in the sqldiff?

File: api/src/main/resources/liquibase.xml [hunk 26]
```diff
@@ -0,0 +1,317 @@
...
+				<constraints nullable="true"/>
+			</column>
+		</addColumn>
+	</changeSet>
+
+	<changeSet id="a5b34ba3-2cb0-46b5-bb47-4e07da4a7acd" author="Jeremy Keiper">
+		<preConditions onFail="MARK_RAN">
+			<columnExists tableName="idgen_seq_id_gen" columnName="max_length"/>
+		</preConditions>
+		<sql>UPDATE `idgen_seq_id_gen` SET `max_length` = `min_length`</sql>
```
Comment (dkayiwa) [inline api/src/main/resources/liquibase.xml]: What happens if min_length does not exist?

File: api/src/main/resources/liquibase.xml (hunk 25 above)
Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: No. Fixed

Comment (segunolalive) [reply api/src/main/resources/liquibase.xml]: Thank you

**Comment by dkayiwa** [comment]
@segunolalive can you also remove the sqldiff file?

**Comment by segunolalive** [comment]
Okay. Done.
